realm/rfc/0001-backlink-query-syntax.md

```diff
@@ -0,0 +1,204 @@
+- Feature Name: Backlink query syntax
+- Start Date: 2016-05-30
+- RFC PR: 
+
+# Summary
+
+It should be possible to define and query backlinks in both `Realm` and 
+`DynamicRealm`. As a `DynamicRealm` can work without _any_ Realm model classes,
+it should be possible to query backlinks without them being defined there.
+
+# Motivation
+
+Backlinks are special in Realm. They are not defined explicitly in the schema,
+but are automatically created when defining Links and LinkLists (forward links).
+
+This creates some conflicts with the current query system as it rely on all 
+fields being defined by the schema. This is called *named backlinks*.
+
+This proposal introduces a new syntax for queries so it is possible to query
+across a backlink even though it hasn't been defined by a model class. This is
+called *unnamed backlinks*.
+
+Any design should ideally solve all these constraints:
+
+- Should support a list of backlinks.
+
+- Should support a single backlink, enforcing the constraint that only 1 object 
+  can point to it. This is not supported by Core, but it is a natural extension.
+
+- Should support backlinks from multiple fields.
+
+- Should be usable from the Dynamic API.
+
+- Should either support polymorphic objects or not prevent them from being 
+  implemented.
+
+- Ideally it should work for both a pure text-based variant as well as the 
+  current mixed style found in Realm Java.
+  
+
+# Detailed design
+
+**Example model classes:**
+
+```
+public class Person extends RealmObject {
+	private Dog favoriteDog;
+	private RealmList<Dog> dogs;
+}
+
+public class Dog extends RealmObject {
+	private String name
+
+	// Named backlinks below
+	@Backlink
+	private RealmResults<Person> owner;
+
+	@Backlink("favoriteDog")
+	private RealmResults<Person> favoriteOwner;
+}
+```
+
+**Introduce the following query extension**
+
+- Use `"y[x]"`to search field Y on type X where X is the backlink type.
+- Use `"y[x.z]"` to search field Y on type X where X is a backlink through 
+  field Z.
+- This is a supplement to defining the named backlinks.
+- No changes to the migration API.
+- No changes to the schema definition.
+
+Example:
+
+```
+// Named backlink queries
+realm.where(Dog.class).equalTo("owners.name", "John").findAll();
+realm.where(Dog.class).equalTo("favoriteOwner.name")
+
+// Unnamed backlink queries
+realm.where(Dog.class).equalTo("name[Person]", "John").findAll();
+realm.where(Dog.class).equalTo("name[Person.favoriteDog]").findAll();
+
+// Example in a purely text based variant
+"name[Person] = 'John'"
+```
```

Comment:
@beeender Do you have another suggestion for making it easier to parse? IMO the suggestion by @astigsen and @emanuelez doesn't really work well because you are adding query methods for the sole benefit of the Dynamic API, which is a deal breaker in my mind.

It is fair to consider the proposed syntax to weird, but would it be learnable? Keep in mind that 
that this is a real corner case, since most people will operate on named backlinks that are defined the model class. Do you have a better suggestion for a String based representation that could work for both the static and dynamic API?

Also, if we believe it will ruin the query language to much, I sketched an alternative in https://github.com/realm/realm-java/pull/2904/files#diff-b684aefe1e063faddcad9b319552f6e2R205 that means we will enforce named backlinks as part of the schema. Is that an acceptable solution instead?